src/middlewares/schema.ts

```diff
@@ -80,6 +100,24 @@ export type ParamSchema = InternalParamSchema & {
       };
 };
 
+type DefaultSchemaKeys = keyof BaseParamSchema | keyof ValidatorsSchema | keyof SanitizersSchema;
+
+/**
+ * Defines a schema of validations/sanitizations for a field
+ */
+export type ParamSchema<T extends string = DefaultSchemaKeys> = BaseParamSchema &
+  ValidatorsSchema &
+  SanitizersSchema &
+  {
+    [K in T]?: K extends keyof BaseParamSchema
+      ? BaseParamSchema[K]
+      : K extends keyof ValidatorsSchema
+      ? ValidatorsSchema[K]
+      : K extends keyof SanitizersSchema
+      ? SanitizersSchema[K]
+      : CustomSanitizerSchemaOptions | CustomValidatorSchemaOptions;
+  };
```

Comment:
Thanks for the ideas!
Can't do much formatting because of prettier 😛 

Don't know if the performance wins will be obvious if I change things around. Maybe it's coincidentally at its optimal already 😄 

The other suggestions are probably not going to work.
`CustomSanitizerSchemaOptions | CustomValidatorSchemaOptions` represents the value really, not a dictionary of possible keys/values that `K` might be in, so I have to dismiss your suggestion :(

But thanks again!!!